From acbcc0c08b2b29dda9da40128d5422138a904844 Mon Sep 17 00:00:00 2001
From: "Yo-Jung (Leo) Lin" <leo.lin@canonical.com>
Date: Wed, 16 Apr 2025 00:00:00 +0000
Subject: [PATCH 1/2] hwmon: (spd5118) pass spd5118_data to hwmon callbacks

Currently only the regcache is passed to the hwmon device. Pass the
whole spd5118_data for better flexibility.

Backported for linux-zen 7.1.2 as used by the viscacha NixOS host.

Link: https://patchew.org/linux/20250416-for-upstream-spd5118-spd-write-prot-detect-v1-1-8b3bcafe9dad@canonical.com/
Signed-off-by: Yo-Jung (Leo) Lin <leo.lin@canonical.com>
[Known Rabbit: adapt context for linux-zen 7.1.2 spd5118.c]
Signed-off-by: Known Rabbit <opensource@rabit.pw>
---
 drivers/hwmon/spd5118.c | 10 +++++++---
 1 file changed, 7 insertions(+), 3 deletions(-)

diff --git a/drivers/hwmon/spd5118.c b/drivers/hwmon/spd5118.c
index 5da44571b6a0ce43b621faba9e67d81491ff0b28..99596538b65969e82bf3c6f1ab0c8134f7f8b753 100644
--- a/drivers/hwmon/spd5118.c
+++ b/drivers/hwmon/spd5118.c
@@ -178,7 +178,8 @@ static int spd5118_read_enable(struct regmap *regmap, long *val)
 static int spd5118_read(struct device *dev, enum hwmon_sensor_types type,
 			u32 attr, int channel, long *val)
 {
-	struct regmap *regmap = dev_get_drvdata(dev);
+	struct spd5118_data *data = dev_get_drvdata(dev);
+	struct regmap *regmap = data->regmap;
 
 	if (type != hwmon_temp)
 		return -EOPNOTSUPP;
@@ -260,7 +261,8 @@ static int spd5118_temp_write(struct regmap *regmap, u32 attr, long val)
 static int spd5118_write(struct device *dev, enum hwmon_sensor_types type,
 			 u32 attr, int channel, long val)
 {
-	struct regmap *regmap = dev_get_drvdata(dev);
+	struct spd5118_data *data = dev_get_drvdata(dev);
+	struct regmap *regmap = data->regmap;
 
 	switch (type) {
 	case hwmon_temp:
@@ -273,6 +275,8 @@ static int spd5118_write(struct device *dev, enum hwmon_sensor_types type,
 static umode_t spd5118_is_visible(const void *_data, enum hwmon_sensor_types type,
 				  u32 attr, int channel)
 {
+	const struct spd5118_data *data = _data;
+
 	if (type != hwmon_temp)
 		return 0;
 
@@ -564,7 +568,7 @@ static int spd5118_common_probe(struct device *dev, struct regmap *regmap,
 	}
 
 	hwmon_dev = devm_hwmon_device_register_with_info(dev, "spd5118",
-							 regmap, &spd5118_chip_info,
+							 data, &spd5118_chip_info,
 							 NULL);
 	if (IS_ERR(hwmon_dev))
 		return PTR_ERR(hwmon_dev);
-- 
2.54.0

